### PR TITLE
Reimplemented polarized Fresnel computation

### PR DESCRIPTION
**Pull request by rbeerwerth**
This PR adds a more stable computational kernel for polarized Fresnel calculations that also supports structural roughness (Tanh + Nevot-Croce). As a precursor, the scalar kernel is also rewritten to be formula-wise identical to the polarized version with different underlying mathematical structures. Key to achieving better numerical precision is to rewrite the transfer-matrix equations in a form partially identical to a Parrat recursion in order to completely avoid the division by the complex phase factor (or matrix in the polarized case).

Intensive testing was performed. For cases where all magnetization is parallel to the spin, comparison to scalar calculations with shifted SLDs shows very good numerical agreement, both for large numbers of layers as well as in the presence of thick layers. In these cases, coefficients and the reflectivity were compared between scalar and polarized computations. Furthermore, the new implementation was tested against the old code, which confirms that the newly introduced coefficient type behaves identical to the old one via its public interface. 

Known problems:
* The matrix RT coefficients do currently have a jump of a factor of two when the limit k0_z -> 0 is considered vs k0_z = 0. This behavior is currently reproduced by the new implementation

### CODE REVIEW

File: Tests/UnitTests/Core/Fresnel/SpecularMagneticTest.cpp [hunk 1]
```diff
@@ -3,46 +3,130 @@
...
+    void test_degenerate();
+
+    void ifEqual(const Eigen::Vector2cd& lhs, const Eigen::Vector2cd& rhs);
 };
 
-TEST_F(SpecularMagneticTest, initial)
+void SpecularMagneticTest::ifEqual(const Eigen::Vector2cd& lhs, const Eigen::Vector2cd& rhs)
+{
+    EXPECT_NEAR(0.0, std::abs(lhs(0) - rhs(0)), eps);
+    EXPECT_NEAR(0.0, std::abs(lhs(1) - rhs(1)), eps);
```
Comment (gpospelov) [inline Tests/UnitTests/Core/Fresnel/SpecularMagneticTest.cpp:37]: I think the previous way of testing was also not bad. Now if you have a failure here, you won't be able to tell which coefficient is failing.

File: Tests/UnitTests/Core/Fresnel/SpecularMagneticTest.cpp [hunk 2]
```diff
@@ -3,46 +3,130 @@
...
+    return std::make_unique<ProcessedSample>(mLayer, SimulationOptions());
 }
 
-TEST_F(SpecularMagneticTest, zerofield)
+TEST_F(SpecularMagneticTest, degenerate)
+{
+    test_degenerate<SpecularMagneticStrategy>();
+}
+
+TEST_F(SpecularMagneticTest, degenerate_new)
```
Comment (gpospelov) [inline Tests/UnitTests/Core/Fresnel/SpecularMagneticTest.cpp:113]: At some point, we could try to find a better way of running tests where everything is the same, except for instance being tested. Here is a nice and short description of [parameterized test](https://www.sandordargo.com/blog/2019/04/24/parameterized-testing-with-gtest) which might be one option.

**Review by gpospelov — APPROVED** [review]

File: Tests/UnitTests/Core/Fresnel/SpecularMagneticOldTest.cpp [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    auto coeffs_zerofield =
+        std::make_unique<SpecularMagneticOldStrategy>()->Execute(sample_zerofield.slices(), k1);
+    MatrixRTCoefficients RTMatrix =
+        *dynamic_cast<const MatrixRTCoefficients*>(coeffs_zerofield[1].get());
+    Eigen::Vector2cd TPM = RTMatrix.T1plus() + RTMatrix.T2plus();
+    Eigen::Vector2cd RPM = RTMatrix.R1plus() + RTMatrix.R2plus();
+    Eigen::Vector2cd TMM = RTMatrix.T1min() + RTMatrix.T2min();
+    Eigen::Vector2cd RMM = RTMatrix.R1min() + RTMatrix.R2min();
+
+    EXPECT_NEAR(0.0, std::abs(TPS(0) - TPM(0)), eps);
```
Comment (jwuttke) [inline Tests/UnitTests/Core/Fresnel/SpecularMagneticOldTest.cpp:76]: Warum nicht einfach `EXPECT_NEAR(TPS(0), TPM(0), eps);`? Kürzer, klarer, und liefert ggf. informativere Fehlermeldungen.

File: Core/Multilayer/SpecularMagneticNewNCStrategy.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+{
+complex_t checkForUnderflow(complex_t val);
+}
+
+std::pair<Eigen::Matrix2cd, Eigen::Matrix2cd>
+SpecularMagneticNewNCStrategy::computeRoughnessMatrices(const MatrixRTCoefficients_v3& coeff_i,
+                                                        const MatrixRTCoefficients_v3& coeff_i1,
+                                                        double sigma) const
+{
+    auto beta_i  = coeff_i.m_lambda(1) - coeff_i.m_lambda(0);
```
Comment (jwuttke) [inline Core/Multilayer/SpecularMagneticNewNCStrategy.cpp:27]: Fürs lesen fände ich es hilfreich, wenn wir `auto` zumindest für die Datentypen `double` und `complex_t` ausbuchstabieren.

File: Core/Multilayer/SpecularMagneticNewNCStrategy.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        QL << (b_p_vec.z() + 1.)/factor1,                (b_p_vec.z() - 1.)/factor2,
+                (b_p_vec.x() + I * b_p_vec.y())/factor1, (b_p_vec.x() + I * b_p_vec.y())/factor2;
+        QR << (b_p_vec.z() + 1.)/factor1,   (b_p_vec.x() - I * b_p_vec.y())/factor1,
+                (b_p_vec.z() - 1.)/factor2, (b_p_vec.x() - I * b_p_vec.y())/factor2;
+
+        auto exp1 = Eigen::Matrix2cd(Eigen::DiagonalMatrix<complex_t, 2>(
+                            {std::exp(alpha_pp), std::exp(alpha_pp)}));
+
+        Eigen::Matrix2cd roughnessMatrix;
+        if( std::abs(beta_p) > std::numeric_limits<double>::epsilon() * 10. )
```
Comment (jwuttke) [inline Core/Multilayer/SpecularMagneticNewNCStrategy.cpp:58]: Bitte `clang-format` über alles laufen lassen.

File: Core/Multilayer/SpecularMagneticNewNCStrategy.cpp [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        {
+            Eigen::Matrix2cd exp2 = Eigen::Matrix2cd(Eigen::DiagonalMatrix<complex_t, 2>(
+                                {std::exp(beta_pp), std::exp(-beta_pp)}));
+            roughnessMatrix = exp1 * QL * exp2 * QR;
```
Comment (jwuttke) [inline Core/Multilayer/SpecularMagneticNewNCStrategy.cpp:62]: hier `return`

File: Core/Multilayer/SpecularMagneticNewNCStrategy.cpp [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+        }else
```
Comment (jwuttke) [inline Core/Multilayer/SpecularMagneticNewNCStrategy.cpp:64]: ... dann braucht's auch kein `else`

File: Core/Multilayer/SpecularMagneticNewStrategy.cpp [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+SpecularMagneticNewStrategy::computeTR(const std::vector<Slice>& slices,
+                                       const std::vector<complex_t>& kzs) const
+{
+    if (slices.size() != kzs.size())
+        throw std::runtime_error(
+            "Error in SpecularMagnetic_::execute: kz vector and slices size shall coinside.");
+    if (slices.empty())
+        return {};
+
+    std::vector<MatrixRTCoefficients_v3> result;
```
Comment (jwuttke) [inline Core/Multilayer/SpecularMagneticNewStrategy.cpp:65]: `_v3` ist zur Umbenennung vorgemerkt?

File: Core/Multilayer/SpecularMagneticNewStrategy.cpp [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+        }
+        return result;
+    }
+
+    calculateUpwards(result, slices);
+
+    return result;
+}
+
+void SpecularMagneticNewStrategy::calculateUpwards(
```
Comment (jwuttke) [inline Core/Multilayer/SpecularMagneticNewStrategy.cpp:97]: Warum wird `coeff` nicht per `return` zurückgeliefert?

File: Core/RT/MatrixRTCoefficients_v3.cpp [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+    auto mat = T1Matrix();
+    return mat * m_T.col(0);
+    ;
+}
+
+Eigen::Vector2cd MatrixRTCoefficients_v3::R1plus() const
+{
+    auto mat = T1Matrix();
+    return mat * m_R.col(0);
+    ;
```
Comment (jwuttke) [inline Core/RT/MatrixRTCoefficients_v3.cpp:90]: s.o.

File: Core/RT/MatrixRTCoefficients_v3.cpp [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+Eigen::Matrix2cd MatrixRTCoefficients_v3::T2Matrix() const
+{
+    return TransformationMatrix(m_lambda(0), {1., 0.});
+}
+
+Eigen::Vector2cd MatrixRTCoefficients_v3::T1plus() const
+{
+    auto mat = T1Matrix();
+    return mat * m_T.col(0);
+    ;
```
Comment (jwuttke) [inline Core/RT/MatrixRTCoefficients_v3.cpp:83]: ?

File: Core/RT/MatrixRTCoefficients_v3.cpp [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+
+Eigen::Matrix2cd MatrixRTCoefficients_v3::T2Matrix() const
+{
+    return TransformationMatrix(m_lambda(0), {1., 0.});
+}
+
+Eigen::Vector2cd MatrixRTCoefficients_v3::T1plus() const
+{
+    auto mat = T1Matrix();
+    return mat * m_T.col(0);
```
Comment (jwuttke) [inline Core/RT/MatrixRTCoefficients_v3.cpp:82]: `return T1Matrix() * m_T.col(0);` fände ich leichter lesbar

File: Core/RT/MatrixRTCoefficients_v3.cpp [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+    const auto exp2 = Eigen::Matrix2cd(Eigen::DiagonalMatrix<complex_t, 2>(
+        {GetImExponential(thickness * m_lambda(1)), GetImExponential(thickness * m_lambda(0))}));
+
+    // Compute resulting phase matrix according to exp(i p_m d_m) = exp1 * Q * exp2 * Q.adjoint();
+    if (std::abs(m_b.mag() - 1.) < eps) {
+        Eigen::Matrix2cd Q;
+        const double factor1 = 2. * (1. + m_b.z());
+        Q << (1. + m_b.z()), (I * m_b.y() - m_b.x()), (m_b.x() + I * m_b.y()), (m_b.z() + 1.);
+
+        result = Q * exp2 * Q.adjoint() / factor1;
```
Comment (jwuttke) [inline Core/RT/MatrixRTCoefficients_v3.cpp:185]: `return` hier

File: Core/RT/MatrixRTCoefficients_v3.cpp [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+
+    } else if (m_b.mag() < eps) {
+        result = Eigen::Matrix2cd::Identity() * GetImExponential(alpha);
```
Comment (jwuttke) [inline Core/RT/MatrixRTCoefficients_v3.cpp:188]: und hier

File: Core/RT/MatrixRTCoefficients_v3.cpp [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+
+    } else
```
Comment (jwuttke) [inline Core/RT/MatrixRTCoefficients_v3.cpp:190]: dann braucht's kein `else`

File: Core/Multilayer/SpecularMagneticNewNCStrategy.cpp [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+{
+    Eigen::Matrix2cd roughness_sum{Eigen::Matrix2cd::Identity()};
+    Eigen::Matrix2cd roughness_diff{Eigen::Matrix2cd::Identity()};
+    if (sigma != 0.) {
+        auto ret = computeRoughnessMatrices(coeff_i, coeff_i1, sigma);
+        roughness_sum = std::get<0>(ret);
+        roughness_diff = std::get<1>(ret);
+    }
+
+    auto P = Eigen::Matrix2cd(coeff_i.computeInverseP() * coeff_i1.computeP());
```
Comment (jwuttke) [inline Core/Multilayer/SpecularMagneticNewNCStrategy.cpp:89]: Hier und an vielen anderen Stellen würde ich ein `const` spendieren

File: Core/Multilayer/SpecularMagneticNewTanhStrategy.cpp [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+#include "Core/Tools/MathFunctions.h"
+
+namespace
+{
+const double pi2_15 = std::pow(M_PI_2, 1.5);
+} // namespace
+
+Eigen::Matrix2cd
+SpecularMagneticNewTanhStrategy::computeRoughnessMatrix(const MatrixRTCoefficients_v3& coeff,
+                                                        double sigma, bool inverse) const
```
Comment (jwuttke) [inline Core/Multilayer/SpecularMagneticNewTanhStrategy.cpp:26]: Diese Funktion verwendet, soweit ich sehe, keine Klassenvariablen und -funktionen? Dann könnte man sie aus der Klasse rausnehmen und in den lokalen anonymen namespace schieben.

**Review by jwuttke — CHANGES_REQUESTED** [review]

**Pull request merged**